lib/ui/painting/fragment_program.cc

```diff
@@ -37,6 +41,71 @@ void FragmentProgram::RegisterNatives(tonic::DartLibraryNatives* natives) {
        FOR_EACH_BINDING(DART_REGISTER_NATIVE)});
 }
 
+void FragmentProgram::initFromAsset(std::string asset_name,
```

Comment:
@jonahwilliams What are the changes to support hot reload? Guessing:
* `FragmentProgram` retains `asset_name`
* On a hot reload, reload the shader data from the asset manager
* Make a new `SkRuntimeEffect` and overwrite `runtime_effect_`
* Also re-setup the fields of the Dart object.
* Other stuff?

The thing I'm not sure how to do is hook into a notification of a hot reload.

Reply:
For step 2, rather than signal on a hot reload (because the reload may not have changed that particular asset)- wait for a specifc vm service notification that contains an invalidation for a particular asset. This would need to be dispatched from the tool, but it could be hooked up alongside the existing vm service listeners in service_protocol.cc .

https://github.com/flutter/engine/blob/e23c43195c6492abcb8bb87fd775d54e0d995947/runtime/service_protocol.cc

Reply:
I'd imagine we'd need to create some sort of shared debug mode only shader manager that could correlate the shader instance with the vm service signal

Reply:
That seems reasonable, also all for doing it in Dart as well!